src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -0,0 +1,376 @@
+        {
+            var symbolEvent = compilationEvent as SymbolDeclaredCompilationEvent;
+            if (symbolEvent != null)
+            {
+                // Add/remove symbol events to per-tree event cache.
+                // Any diagnostics request for a tree should trigger symbol and syntax node analysis for symbols with at least one declaring reference in the tree.
+                foreach (var location in symbolEvent.Symbol.Locations)
+                {
+                    if (location.SourceTree != null)
+                    {
+                        if (generated)
+                        {
+                            _eventsCache.AddOrUpdate(location.SourceTree, ImmutableHashSet.Create(compilationEvent), (tree, events) => events.Add(compilationEvent));
+                        }
+                        else
+                        {
+                            RemoveCompilationEvent(location.SourceTree, compilationEvent);
+                        }
+                    }
+                }
+            }
+            else
+            {
+                // Add/remove compilation unit completed events to per-tree event cache.
+                var compilationUnitCompletedEvent = compilationEvent as CompilationUnitCompletedEvent;
+                if (compilationUnitCompletedEvent != null)
+                {
+                    var tree = compilationUnitCompletedEvent.SemanticModel.SyntaxTree;
+                    if (generated)
+                    {
+                        _eventsCache.AddOrUpdate(tree, ImmutableHashSet.Create(compilationEvent), (t, events) => events.Add(compilationEvent));
+                    }
+                    else
+                    {
+                        RemoveCompilationEvent(tree, compilationEvent);
+                    }
+                }
+                else if (compilationEvent is CompilationStartedEvent || compilationEvent is CompilationCompletedEvent)
+                {
+                    // Add/remove compilation events to '_nonSourceEvents' event cache.
+                    if (generated)
+                    {
+                        _nonSourceEvents.Add(compilationEvent);
+                    }
+                    else
+                    {
+                        _nonSourceEvents.Remove(compilationEvent);
+                    }
+                }
+                else
+                {
+                    throw new InvalidOperationException("Unexpected compilation event of type " + compilationEvent.GetType().Name);
+                }
+            }
+        }
+
+        private void RemoveCompilationEvent(SyntaxTree tree, CompilationEvent compilationEvent)
+        {
+            ImmutableHashSet<CompilationEvent> currentEvents;
+            if (_eventsCache.TryGetValue(tree, out currentEvents) && currentEvents.Contains(compilationEvent))
+            {
+                _eventsCache.TryUpdate(tree, currentEvents.Remove(compilationEvent), currentEvents);
```

Comment:
hmm.. so if event has changed, already processed compilation event never removed from set? by looking at this code, immutable set might not a good choice? rather just regular concurrent set might better?